src/rust/engine/fs/src/directory.rs

```diff
@@ -311,6 +314,43 @@ impl From<&File> for remexec::FileNode {
   }
 }
 
+#[derive(Clone, Debug, DeepSizeOf)]
+pub struct Symlink {
+  name: Name,
+  target: String,
+}
+
+impl Symlink {
+  pub fn name(&self) -> Name {
+    self.name
+  }
+
+  pub fn target(&self) -> String {
+    self.target.to_owned()
+  }
+}
+
+impl TryFrom<&remexec::SymlinkNode> for Symlink {
+  type Error = String;
+
+  fn try_from(symlink_node: &remexec::SymlinkNode) -> Result<Self, Self::Error> {
+    Ok(Self {
+      name: Name(Intern::from(&symlink_node.name)),
+      target: symlink_node.target.clone(),
+    })
+  }
+}
+
+impl From<&Symlink> for remexec::SymlinkNode {
+  fn from(symlink: &Symlink) -> Self {
+    remexec::SymlinkNode {
+      name: symlink.name.as_ref().to_owned(),
+      target: symlink.target.clone(),
+      ..remexec::SymlinkNode::default()
+    }
+  }
+}
+
 // TODO: `PathStat` owns its path, which means it can't be used via recursive slicing. See
 // whether these types can be merged.
 enum TypedPath<'a> {
```

Comment:
The fact that this code hasn't changed suggests that `PathStat`, which is the type produced while `VFS::expand` walks the filesystem has not been changed.

If you get tests passing without that change, then I _think_ that what you will have accomplished is that our VFS code (which is what matches globs in the repo for `PathGlobs`->`Digest`, and which [captures the outputs of processes from sandboxes](https://github.com/pantsbuild/pants/blob/f72b8c0366aba7fa0a16bb83d49983113e8e44af/src/rust/engine/process_execution/src/local.rs#L491-L518)) will never actually create a `DigestTrie`/`Digest`/`Snapshot` containing a symlink, BUT when one was created synthetically, it would present as such in a memory, and be successfully materialized.

That's probably a good breaking point for a first PR, so you might want to avoid dipping into `VFS` yet if you can.

Reply:
Yeah just started to muck with `fs_test` and noticed this behavior.

I like the idea of splitting into multiple PRs.